Pull request for #3: Stata package installation

M-R-Sullivan opened the pull request:
For #3. I used [this repo](https://github.com/sergiocorreia/stata-misc) as a model for allowing installation directly from GitHub via the Stata console. 

You should be able to install the repo's packages by following the readme instructions, although you'll need to substitute `issue3-install_packages` for `master` in the paths to GitHub.

Let me know if you think more extensive attribution/licensing is warranted.

arosenbe commented on crosswalk/ado/stata.toc:
```diff
@@ -0,0 +1,3 @@
+p make_input_template Write an input template
```
Did Matt and Jesse write this? If so, we should give them attribution as for xtreg_cg.

M-R-Sullivan replied:
It wasn't clear to me whether they did or not.

arosenbe commented on gslab_misc/ado/stata.toc:
```diff
@@ -0,0 +1,34 @@
+p aivreg                Estimates panel IV models using 2SLS within estimator
```
Maybe mention that this is a collection of third part packages and provide attribution in the various .pkg files.

M-R-Sullivan replied:
- I added author attributions
- All of the ado files except two either had no authorship label or had help files lising MG, JMS, and/or RAs as authors. Those two are:
    - matrix_to_txt.ado: by  Michael Blasnik (M. Blasnik & Associates, Boston, MA), Ben Jann (ETH Zurich), and Matthew Gentzkow
    - placevar.ado: by Jeroen Weesie (Dept of Sociology/ICS, Utrecht University)

arosenbe commented on xtreg_ch/ado/xtreg_ch_loglik.pkg:
```diff
@@ -0,0 +1,6 @@
+d
+d xtreg_ch_loglik: log likelihood for xtreg_ch
+d
+f xtreg_ch_loglik.ado
```
No help?

M-R-Sullivan replied:
There wasn't one before.

arosenbe commented on xtreg_ch/readme.md:
```diff
@@ -1,2 +1,13 @@
-xtreg_ch runs a random-effects regression allowing for parametric correlation between RHS variables and random effect as well as heteroskedasticity in individual errors.
+xtreg_ch runs a random-effects regression allowing for parametric correlation 
+between RHS variables and random effect as well as heteroskedasticity in 
+individual errors.
```
Why the line breaks?

M-R-Sullivan replied:
Ease of reading.

M-R-Sullivan commented on gslab_misc/ado/cf_mg.pkg:
```diff
@@ -0,0 +1,6 @@
+d
+d cf_mg: Like -cf-, but the exit flag is always zero
+d
+f cf_mg.ado
```
There wasn't a hlp file in the repo before.

M-R-Sullivan commented on gslab_misc/ado/plotcoeffs_nolab.pkg:
```diff
@@ -0,0 +1,6 @@
+d
+d plotcoeffs_nolab: Plot coefficients after a regression (no labels)
+d
+f plotcoeffs_nolab.ado
```
There wasn't a help file in the repo, I believe.

M-R-Sullivan commented on gslab_misc/ado/cutby.pkg:
```diff
@@ -0,0 +1,7 @@
+d
+d cutby: By-able version of egen ... = cut()
+d
+f cutby.hlp
```
Good catch!

M-R-Sullivan commented:
Note to self: get all RAs to reinstall `gslab_stata` once this passes PR. 

📊

M-R-Sullivan commented on gslab_misc/ado/checkdta.pkg:
```diff
@@ -0,0 +1,8 @@
+d gslab_stata
+d
+d checkdta: Create checkdta.log, which allows users to check whether Stata
+datasets have actually changed after make.bat is re-run.
```
No

M-R-Sullivan commented on gslab_misc/ado/cf_mg.pkg:
```diff
@@ -0,0 +1,6 @@
+d gslab_stata
+d
+d cf_mg: Like -cf-, but the exit flag is always zero
```
I will keep as is.

M-R-Sullivan commented on gslab_misc/ado/load_and_append.pkg:
```diff
@@ -0,0 +1,7 @@
+d gslab_stata
+d
+d load_and_append: Load_and_append appends a group of files and loads the appended file into memory.
```
I will standardize

M-R-Sullivan commented on gslab_misc/ado/placevar.pkg:
```diff
@@ -0,0 +1,6 @@
+d gslab_stata
+d
+d placevar: Orders varlist
+f placevar.ado
```
Fixed

M-R-Sullivan commented on gslab_misc/ado/rankunique.pkg:
```diff
@@ -0,0 +1,7 @@
+d gslab_stata
+d
+d rankunique: Replacement for egen ... = rank(), unique that uses explicit randomization
```
OK.

M-R-Sullivan commented:
Still working on this.

M-R-Sullivan commented:
@gentzkow and @jmshapir:

We noticed the following things when preparing this repo for its public release.
- I thought @arosenbe's suggestion that we give this repo an MIT license was a good one, but I noticed that two scripts in `gslab_misc` have authors other than either of you or former RAs. Namely, (i)  Michael Blasnik (M. Blasnik & Associates, Boston, MA) and Ben Jann (ETH Zurich) are listed among the authors of `matrix_to_txt`, and (ii) Jeroen Weesie (Dept of Sociology/ICS, Utrecht University) is the author of `placevar`. If you don't think we should put these files in the licensed repo, we could remove them or reconsider making the repo public. Also,  there are some other ado files without listed authors.

- A few of the unit tests for this repo's ado files fail. Do you think it is worth spending time to understand why and implement fixes as a part of this pull request, or would you like to proceed in some other way?

jmshapir commented:
@M-R-Sullivan we should see what @gentzkow thinks but here are my reactions:
* If those ADO files are literally from the outside world, then they probably shouldn't be in ```gslab-misc``` anyway.  If we want to install third-party tools as part of our setup, we should probably use [ssc](http://www.stata.com/support/ssc-installation/). Certainly we do not want to try to license anything we did not write.
* I would venture it's worth an hour or so of work to debug unit tests, not more. But we can see what @gentzkow thinks. I am happy to defer to him.

M-R-Sullivan commented:
@arosenbe - thanks for your helpful comments! I acted on most of your suggestions. There are a few things left to do that I was hoping you could wrap up. Some notes:
- I added an SVN-free testing script to `gslab_misc`.  If you dislike my approach, you can change it here or in a follow-up issue. If you like it, you can add it to the other packages here or in a follow-up issue.
- I added author attributions but not versions to the package files. You/another lab member can add them if you think they'd provide some benefit.
- Per our Skype discussion, the provenance of some of the `gslab_misc` packages is unclear. It is probably best to remove all of those that we aren't sure we wrote before adding a license. I initially made this issue because we needed a way of getting some (maybe just `matrix_to_txt`) of these ado files for `divergence`. So you should probably make a note to update the `divergence` documentation with the new instructions for installing the required ado files.

Let me know if I can help out with anything here!

lboxell commented:
@M-R-Sullivan,

Quick question: Did you ever successfully `net install` from the current branch?

M-R-Sullivan commented:
Indeed I did!

On Jul 11, 2017 6:09 PM, "Levi Boxell" <notifications@github.com> wrote:

> @M-R-Sullivan <https://github.com/m-r-sullivan>,
>
> Quick question: Did you ever successfully net install from the current
> branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gslab-econ/gslab_stata/pull/4#issuecomment-314564624>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATFIO0dNz81NV_XrbRrSiB1an17iTeqPks5sM91_gaJpZM4NtlM2>
> .
>

lboxell commented:
@arosenbe,

The net installation works (I needed to delete the `/tree` part of the URL), and I've removed the packages that MG didn't confirm we created (with the note regarding mat2txt included). 

I haven't included a license at the top level, as I'm not sure what the best way to handle this is with the matrix_to_txt function. 

I'm inclined to make a decision regarding the license and wrap this up, since the primary purpose was to just get proper download instructions.

arosenbe commented:
@lboxell, I'd read [this summary by Nick Cox](https://www.stata.com/statalist/archive/2004-03/msg00840.html), add a license at the top level with plenty of acknowledgement for mat2txt's original authors (noting specifically what they did and did not do), and call it a day.

lboxell commented:
@arosenbe,

See my commit above for license/readme changes. Let me know if you have any other comments or if I can go ahead and squash and merge.

arosenbe commented:
Good to go @lboxell. it's about time we wrapped this up.
